applications/FluidDynamicsApplication/custom_processes/mass_conservation_check_process.cpp

```diff
@@ -0,0 +1,189 @@
+#include "processes/find_nodal_h_process.h"
+
+// Application includes
+#include "mass_conservation_check_process.h"
+
+
+namespace Kratos
+{
+
+/* Public functions *******************************************************/
+
+MassConservationCheckProcess::MassConservationCheckProcess(
+        ModelPart& rModelPart, 
+        const int MassComputationFreq,
+        const bool CompareToInitial, 
+        const bool WriteToLogFile)
+    : Process(), mrModelPart(rModelPart) {
+
+    mMassComputationFreq = MassComputationFreq;
+    mCompareToInitial = CompareToInitial;
+    mWriteToLogFile = WriteToLogFile;
+}
+
+
+MassConservationCheckProcess::MassConservationCheckProcess(
+    ModelPart& rModelPart,
+    Parameters& rParameters)
+    : Process(), mrModelPart(rModelPart) {
+
+    Parameters default_parameters( R"(
+    {
+        "model_part_name"                        : "default_model_part_name",
+        "mass_computation_frequency"             : 20,
+        "compare_to_initial_values"              : true,
+        "write_to_log_file"                      : true
+    }  )" );
+
+    rParameters.ValidateAndAssignDefaults(default_parameters);
+
+    mMassComputationFreq = rParameters["mass_computation_frequency"].GetInt();
+    mCompareToInitial = rParameters["compare_to_initial_values"].GetBool();
+    mWriteToLogFile = rParameters["write_to_log_file"].GetBool();
+}
+
+
+bool MassConservationCheckProcess::GetUpdateStatus(){
+
+    const int time_step = mrModelPart.GetProcessInfo()[STEP];
+
+    if ( mCompareToInitial && time_step == 1){
+        // getting initial values and storing as reference
+        this->ComputeVolumesOfFluids( mInitialPositiveVolume, mInitialNegativeVolume );
+    }
+
+    if ( time_step % this->mMassComputationFreq == 0){
+        
+        // writing an output at a given frequncy
+        this->ComputeVolumesOfFluids( mCurrentPositiveVolume, mCurrentNegativeVolume );
+
+#ifdef KRATOS_DEBUG
+        std::cout << " --- Volume Checking Process --- " << std::endl;
+        std::cout << " - positive Volume = " << mCurrentPositiveVolume << std::endl;
+        if ( mCompareToInitial ){ 
+            std::cout << "   ( " << mCurrentPositiveVolume / mInitialPositiveVolume * 100.0 << "% of initial )" << std::endl; 
+        }
+        std::cout << " - negative Volume = " << mCurrentNegativeVolume << std::endl;
+        if ( mCompareToInitial ){ 
+            std::cout << "   ( " << mCurrentNegativeVolume / mInitialNegativeVolume * 100.0 << "% of initial )" << std::endl; 
+        }
+        std::cout << " --- --- --- --- --- --- --- --- " << std::endl;
+#endif
+        mIsUpdated = true;
+        return true;
+
+    } else {
+
+        mCurrentPositiveVolume = -1.0;
+        mCurrentNegativeVolume = -1.0;
+        return false;
+    }
+};
+
+
+/* Private functions ****************************************************/
+
+void MassConservationCheckProcess::ComputeVolumesOfFluids( double &positiveVolume, double &negativeVolume ){
```

Comment:
do you need the arguments to be references?
If not then pass the standard-datatypes by-value

Reply:
Pass-by-value was implemented

Reply:
I am afraid, I do need to pass-by-reference, because these are output paramters. Is that ok?

Reply:
sure, if you need it it is ok

If others read/want to understand your code and they see a reference, they would assume that there is something special going on, which is the case for you